File: daily_problems/2024/08/0801/personal_submission/cf196b_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    vector<string> grid(n);
    for (int i = 0; i < n; i ++)
        cin >> grid[i];
    
    int sx, sy;
    for (int i = 0; i < n; i ++)
        for (int j = 0; j < m; j ++)
            if (grid[i][j] == 'S')
                sx = i, sy = j;

    int inf = 1e9;
    vector<vector<pair<int, int>>> vis(n, vector<pair<int, int>>(m, {inf, inf}));
    vector<pair<int, int>> dirs = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    bool flg = false;

    auto dfs = [&] (auto &&dfs, int x, int y) -> bool {
        if (flg) return true;
        for (auto &[dx, dy]: dirs) {
            int nx = x + dx, ny = y + dy;
            int px = (nx % n + n) % n, py = (ny % m + m) % m;
            if (grid[px][py] == '#') continue;
            if (vis[px][py].first == inf) {
                vis[px][py] = {nx, ny};
                if (dfs(dfs, nx, ny))
                    return true;
            }
            else if (vis[px][py].first != nx || vis[px][py].second != ny) {
                flg = true;
                return true;
            }
        }
        return false;
    };

    vis[sx][sy] = {sx, sy};
    dfs(dfs, sx, sy);

    cout << (flg ? "Yes" : "No");

    return 0;
}

File: daily_problems/2024/08/0801/personal_submission/cf196b_yawn_sean.py
# Submission link: https://codeforces.com/contest/196/submission/273791205
def main():
    n, m = MII()
    grid = [I() for _ in range(n)]

    def f(x, y):
        return x * m + y

    vis = [None] * (n * m)
    sx, sy = 0, 0

    for i in range(n):
        for j in range(m):
            if grid[i][j] == 'S':
                sx = i
                sy = j

    stk = [(sx, sy)]
    vis[f(sx, sy)] = (sx, sy)

    dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    while stk:
        x, y = stk.pop()
        for dx, dy in dirs:
            nx = x + dx
            ny = y + dy
            px = nx % n
            py = ny % m
            if grid[px][py] != '#':
                if vis[f(px, py)] is not None:
                    if vis[f(px, py)] != (nx, ny):
                        print('Yes')
                        exit()
                else:
                    vis[f(px, py)] = (nx, ny)
                    stk.append((nx, ny))

    print('No')